model/__init__.py

```diff
@@ -0,0 +1,533 @@
+    created = getattr(model, create_method, model)(**kwargs)
+    db.add(created)
+    flush(db)
+    return created, True
+
+def get_one(db, model, on_multiple='error', constraint=None, **kwargs):
+    """Gets an object from the database based on its attributes.
+
+    :param constraint: A single clause that can be passed into
+        `sqlalchemy.Query.filter` to limit the object that is returned.
+    :return: object or None
+    """
+    constraint = constraint
+    if 'constraint' in kwargs:
+        constraint = kwargs['constraint']
+        del kwargs['constraint']
+
+    q = db.query(model).filter_by(**kwargs)
+    if constraint is not None:
+        q = q.filter(constraint)
+
+    try:
+        return q.one()
+    except MultipleResultsFound, e:
+        if on_multiple == 'error':
+            raise e
+        elif on_multiple == 'interchangeable':
+            # These records are interchangeable so we can use
+            # whichever one we want.
+            #
+            # This may be a sign of a problem somewhere else. A
+            # database-level constraint might be useful.
+            q = q.limit(1)
+            return q.one()
+    except NoResultFound:
+        return None
+
+def get_one_or_create(db, model, create_method='',
+                      create_method_kwargs=None,
+                      **kwargs):
+    one = get_one(db, model, **kwargs)
+    if one:
+        return one, False
+    else:
+        __transaction = db.begin_nested()
+        try:
+            # These kwargs are supported by get_one() but not by create().
+            get_one_keys = ['on_multiple', 'constraint']
+            for key in get_one_keys:
+                if key in kwargs:
+                    del kwargs[key]
+            obj = create(db, model, create_method, create_method_kwargs, **kwargs)
+            __transaction.commit()
+            return obj
+        except IntegrityError, e:
+            logging.info(
+                "INTEGRITY ERROR on %r %r, %r: %r", model, create_method_kwargs,
+                kwargs, e)
+            __transaction.rollback()
+            return db.query(model).filter_by(**kwargs).one(), False
+
+def numericrange_to_string(r):
+    """Helper method to convert a NumericRange to a human-readable string."""
+    if not r:
+        return ""
+    lower = r.lower
+    upper = r.upper
+    if upper is None and lower is None:
+        return ""
+    if lower and upper is None:
+        return str(lower)
+    if upper and lower is None:
+        return str(upper)
+    if not r.upper_inc:
+        upper -= 1
+    if not r.lower_inc:
+        lower += 1
+    if upper == lower:
+        return str(lower)
+    return "%s-%s" % (lower,upper)
+
+def numericrange_to_tuple(r):
+    """Helper method to normalize NumericRange into a tuple."""
+    if r is None:
+        return (None, None)
+    lower = r.lower
+    upper = r.upper
+    if lower and not r.lower_inc:
+        lower -= 1
```

Comment:
@leonardr wasn't this a bug you fixed?

Reply:
Yeah, but I fixed it as a side thing in https://github.com/NYPL-Simplified/server_core/pull/942, which hasn't been merged yet. (Looking at that PR, I see I also broke out the list of media out of the definition of Edition.medium two different ways.) I'll resolve it in whichever branch gets merged second.